### PR TITLE
exportmanager: add export manifest

### PR DESCRIPTION
**Pull request by mvo5**
This commit adds the first part of the new export manager. It just
adds the `Manifest` datatype and the handling of exporting and
unexporting files and the special handling for the snapd tooling.

The goal is that eventually there will be a full manager that
is responsible for exporting parts of snaps into a new
```
/var/lib/snapd/export/<snapName>/<exportedVersion>/<exportSet>
```
location. The main use case for this is to export the snapd
tooling to a fixed location
```
/var/lib/snapd/export/snapd/current/tools
```
and snap-confine will the use these tools from the new location.

This is split from https://github.com/snapcore/snapd/pull/9546
and slightly simplified.

Ref: https://forum.snapcraft.io/t/12139

### CODE REVIEW

**Comment by pedronis** [comment]
@mvo5 there's a conflict here

File: overlord/exportstate/special.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+// than snapd that need consideration. The result is, in order of preference:
+//
+// 0) "host" if on classic with disabled re-execution.
+// 1) snapd version, if available
+// 2) core version, if available
+// 3) "host" version, if on classic
+//
+// If no provider is available and there is nothing to export the returned
+// version is empty.
+func effectiveExportedVersionForSnapdOrCore(st *state.State) (exportedVersion string, err error) {
```
Comment (pedronis) [inline overlord/exportstate/special.go:122]: I'm not sure if defining this here is premature or not

File: overlord/exportstate/manifest.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+	// Sets describe groups of exported files.
+	Sets map[string]ExportSet `json:"sets"`
+}
+
+// ExportSet describes a group of files for a given type of consumer.
+type ExportSet struct {
+	Name string `json:"name"`
+	// ConsumerIsHost is true if an export set provides files usable from the host system.
+	ConsumerIsHost bool                    `json:"consumer-is-host,omitempty"`
```
Comment (pedronis) [inline overlord/exportstate/manifest.go:50]: we probably need to chat, there is something not anymore completely clear to me about why we decide this at this level

**Review by pedronis — COMMENTED** [review]
did a quick pass  again, I have question that we need to discuss

File: overlord/exportstate/special.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+	if release.OnClassic && os.Getenv("SNAP_REEXEC") == "0" {
```
Comment (pedronis) [inline overlord/exportstate/special.go:123]: there was quite a bit of discussion around this in the original PR

Comment (pedronis) [reply overlord/exportstate/special.go:123]: I'm referrring to this:

https://github.com/snapcore/snapd/pull/9546/files#r525943883

File: overlord/exportstate/manifest.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+)
+
+// Manifest describes content content exported to snaps or the host.
+type Manifest struct {
+	SnapInstanceName          string        `json:"snap-instance-name"`
+	SnapRevision              snap.Revision `json:"snap-revision"`
+	ExportedForSnapdAsVersion string        `json:"exported-for-snapd-as-version,omitempty"`
```
Comment (mardy) [inline overlord/exportstate/manifest.go:35]: A comment on this field would be helpful (to me, at least :-) )

File: overlord/exportstate/manifest.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+	// SourceIsHost is only true if provider of the manifest is not a snap but the classic system.
+	// All SourcePath fields, as visible through Sets[*].Exports[*].SourcePath, are absolute names
+	// in the host file system.
+	SourceIsHost bool `json:"source-is-host,omitempty"`
+
+	// Sets describe groups of exported files.
+	Sets map[string]ExportSet `json:"sets"`
```
Comment (mardy) [inline overlord/exportstate/manifest.go:43]: In this map (as well in the `Exports` field of `ExportSet` below) it's not clear what we use as keys. Do we use the `Name` field of the value?

File: overlord/exportstate/manifest.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+// NewManifestForSnap returns the manifest of a given snap or nil if there is
+// no manifest.
+//
+// Currently only the core and snapd snaps export any content to the system. As
+// such the export manifest is not embedded into the snap meta-data but instead
+// computed here.
+//
+// Both snapd and core snaps export content under the snap name "snapd", using
+// the export set name "tools". The revision is mangled, for "snapd" it is used
+// directly. For "core" it is transformed to "core_$revision".
```
Comment (mardy) [inline overlord/exportstate/manifest.go:76]: I'm not sure whether this is the best place for this comment; at least, I cannot relate it with the function below.

Edit: after looking at the rest of the code, I suggest moving it above, where you define the Manifest type and specifically the `ExportedForSnapdAsVersion` field.

File: overlord/exportstate/manifest.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+		snapInstanceName = "snapd"
+		snapRevision = manifest.ExportedForSnapdAsVersion
+	}
+	return filepath.Join(dirs.ExportDir, snapInstanceName, snapRevision, set.Name, exported.Name)
+}
+
+// exportedFileSourcePath returns the source path that is to be exported.
+func exportedFileSourcePath(manifest *Manifest, set *ExportSet, exported *ExportedFile) string {
+	// Consumer uses host mount namespace
+	if set.ConsumerIsHost {
```
Comment (mardy) [inline overlord/exportstate/manifest.go:153]: super nitpick: not sure if this is just me, but I would find it more readable if such comments where *inside* the `if` block. If you really want to keep it before the `if`, then I'd rephrase it to "Check if the consumer uses host mount namespace".

The other comments inside this `if` block are correctly placed, OTOH :-)

File: overlord/exportstate/manifest.go [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+		if manifest.SourceIsHost {
+			// host-to-host, no translation needed.
+			return exported.SourcePath
+		}
+		// snap-to-host
+		return filepath.Join(dirs.SnapMountDir, manifest.SnapInstanceName, manifest.SnapRevision.String(), exported.SourcePath)
+	}
+	// Consumer uses snap mount namespace
+	if manifest.SourceIsHost {
```
Comment (mardy) [inline overlord/exportstate/manifest.go:162]: Same here, I'd swap these two lines

File: overlord/exportstate/manifest.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+// exportedFileSourcePath returns the source path that is to be exported.
+func exportedFileSourcePath(manifest *Manifest, set *ExportSet, exported *ExportedFile) string {
+	// Consumer uses host mount namespace
+	if set.ConsumerIsHost {
+		if manifest.SourceIsHost {
+			// host-to-host, no translation needed.
+			return exported.SourcePath
+		}
+		// snap-to-host
+		return filepath.Join(dirs.SnapMountDir, manifest.SnapInstanceName, manifest.SnapRevision.String(), exported.SourcePath)
```
Comment (mardy) [inline overlord/exportstate/manifest.go:159]: I probably need a picture to visualize this :-) , but don't we need to do the same check on `manifest.ExportedForSnapdAsVersion` here (and in the last line of this function)?

File: overlord/exportstate/manifest.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+	pathName := exportedFilePath(manifest, set, exported)
+	if err := os.Remove(pathName); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	// Chomp three levels up: once for export set name, once for exported
+	// version (or revision) and finally once more for exported snap name (or
+	// snapd). This way we do not need to repeat the special case described in
+	// exportedFilePath.
+	dir := filepath.Dir(pathName)
+	for i := 0; i < 3; i++ {
```
Comment (mardy) [inline overlord/exportstate/manifest.go:204]: I can imagine this going terribly wrong if we are not super careful :-)
Could we instead compute what is the base path where we want to stop, and change the `for` loop condition to `dir != basePath`? (of course, we should also check that `pathName` starts with `basePath`, otherwise this is even more dangerous :-) )

File: overlord/exportstate/special.go [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+	}
+}
+
+// manifestForCoreSnap returns the manifest of the core snap.
+func manifestForCoreSnap(info *snap.Info) *Manifest {
+	tools := exportSetSnapdTools("/usr/lib/snapd")
+	return &Manifest{
+		SnapInstanceName:          info.InstanceName(),
+		SnapRevision:              info.Revision,
+		ExportedForSnapdAsVersion: fmt.Sprintf("core_%s", info.Revision), // Exception from the rule
```
Comment (mardy) [inline overlord/exportstate/special.go:103]: it's still not clear to me why we do this translation and don't let this export to be named after the `core` snap itself, instead of exposing it as `snapd`.

**Review by mardy — COMMENTED** [review]
The code looks good to me, but I've written a few inline comments where the comment could be expanded a bit.

File: overlord/exportstate/special.go (hunk 3 above)
Comment (mvo5) [reply overlord/exportstate/special.go:123]: Thanks, I see - I moved this to use snapdtool.IsReexec() now but given that having this here might be premature I will probably split it into the next PR instead.

File: overlord/exportstate/manifest.go (hunk 2 above)
Comment (mvo5) [reply overlord/exportstate/manifest.go:50]: I think we can remove this entirely from this intiial PR? AIUI this not needed for the initial MVP of "let's export the snapd tooling"?

File: overlord/exportstate/manifest.go (hunk 5 above)
Comment (mardy) [reply overlord/exportstate/manifest.go:43]: Ansering myself: yes, it looks like we do :-)

File: overlord/exportstate/manifest.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+// Manifest describes content content exported to snaps or the host.
+type Manifest struct {
+	SnapInstanceName          string        `json:"snap-instance-name"`
+	SnapRevision              snap.Revision `json:"snap-revision"`
+	ExportedForSnapdAsVersion string        `json:"exported-for-snapd-as-version,omitempty"`
+
+	// SourceIsHost is only true if provider of the manifest is not a snap but the classic system.
+	// All SourcePath fields, as visible through Sets[*].Exports[*].SourcePath, are absolute names
+	// in the host file system.
+	SourceIsHost bool `json:"source-is-host,omitempty"`
```
Comment (mardy) [inline overlord/exportstate/manifest.go:40]: This seems redundant, because it seems to me that this will be set if and only if `SnapInstanceName` and `SnapRevision` are empty; so, the information is already conveyed by those two fields (or even just by the first one).

File: overlord/exportstate/manifest.go (hunk 4 above)
Comment (mardy) [reply overlord/exportstate/manifest.go:35]: If one of the goals is to support GPU drivers, then we might need to be able to export the content into other snap names than just `snapd` (since I guess that GPU drivers should not appear as being published by snapd?).

So, I suggest to replace this field with a couple of fields:
```
ExportedAsSnapInstanceName string
ExportedAsSnapVersion string
```
Then indeed for the time being we'll have a check that `ExportedAsSnapInstanceName` can only be equal to `snapd` for now, but it's better if we plan the manifest to be generic.

File: overlord/exportstate/special.go (hunk 11 above)
Comment (mardy) [reply overlord/exportstate/special.go:94]: I eventually got it :-)

**Review by mardy — COMMENTED** [review]
I've started looking at this again, and I've added some comments. I will prepare a branch of mine with some changes, as @mvo5 told me he's unlikely to find the time to push this forward.

File: overlord/exportstate/manifest.go (hunk 4 above)
Comment (mardy) [reply overlord/exportstate/manifest.go:35]: After some more thinking: wouldn't it be better to move these fields into the `ExportSet` type? In the future the host system might export not only snapd tools, but also the GPU drivers under another snap name.

**Comment by mvo5** [comment]
Not actionable right now sadly.
